### ISSUE TITLE
Compaction-Planner worker thread dies during compaction

### ISSUE THREAD

**Issue by Manno15** [apache/accumulo#2362]
**Describe the bug**
Compaction planner threads are dying with an IllegalArgumentException during an internal compaction. 

<details>

```
[threads.AccumuloUncaughtExceptionHandler] ERROR: Caught an Exception in Thread[CompactionPlanner-Worker-35,5,main]. Thread is dead.
java.lang.IllegalArgumentException: Unknown runnable type io.opentelemetry.context.Context$$Lambda$243/0x00000008403f8040
        at org.apache.accumulo.tserver.compactions.InternalCompactionExecutor.getJob(InternalCompactionExecutor.java:152) ~[accumulo-tserver-2.1.0-SNAPSHOT.jar:2.1.0-SNAPSHOT]
        at java.util.Comparator.lambda$comparing$ea9a8b3a$1(Comparator.java:436) ~[?:?]
        at java.util.concurrent.PriorityBlockingQueue.siftUpUsingComparator(PriorityBlockingQueue.java:377) ~[?:?]
        at java.util.concurrent.PriorityBlockingQueue.offer(PriorityBlockingQueue.java:488) ~[?:?]
        at java.util.concurrent.ThreadPoolExecutor.execute(ThreadPoolExecutor.java:1347) ~[?:?]
        at org.apache.accumulo.core.util.threads.ThreadPools$1.execute(ThreadPools.java:186) ~[accumulo-core-2.1.0-SNAPSHOT.jar:2.1.0-SNAPSHOT]
        at org.apache.accumulo.tserver.compactions.InternalCompactionExecutor.submit(InternalCompactionExecutor.java:180) ~[accumulo-tserver-2.1.0-SNAPSHOT.jar:2.1.0-SNAPSHOT]
        at org.apache.accumulo.tserver.compactions.CompactionService.submitCompactionJob(CompactionService.java:369) ~[accumulo-tserver-2.1.0-SNAPSHOT.jar:2.1.0-SNAPSHOT]
        at org.apache.accumulo.tserver.compactions.CompactionService.lambda$submitCompaction$4(CompactionService.java:247) ~[accumulo-tserver-2.1.0-SNAPSHOT.jar:2.1.0-SNAPSHOT]
        at io.opentelemetry.context.Context.lambda$wrap$1(Context.java:207) ~[opentelemetry-context-1.7.1.jar:1.7.1]
        at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1128) ~[?:?]
        at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:628) ~[?:?]
        at io.opentelemetry.context.Context.lambda$wrap$1(Context.java:207) ~[opentelemetry-context-1.7.1.jar:1.7.1]
        at java.lang.Thread.run(Thread.java:829) [?:?]
```
</details

**Versions :**
 - Affected version: Current head of main (f8bb900ae080fe0f54dfe04f9e1ad8c4dd2e7930 as of writing this). Though I speculate issue may be due to changes in b6a466348807fe0d23203f3776cbc3119d5a037d.

**To Reproduce**
Steps to reproduce the behavior
1. Setup an Accumulo instance
2. Create a table utilizing the new compaction server props (I used accumulo-testing for this)
3. Use accumulo-testing to continuously-ingest data
4. After a short time, the exception gets thrown. (Seems a manual compaction will throw an exception as well)

**Additional context**
The error message seems to point to it being related to the recent open-telemetry addition in #2259.

**Comment by ctubbsii** [comment]
I think I have a fix for this. We're using instanceof to enforce that the BlockingQueue only contains InternalJob types, and not other Runnables. However, I think we can do this at compile time with better generics instead of waiting until runtime.

**Comment by ctubbsii** [comment]
After some thought, I realized my first attempt won't work, but I think I can still fix this.

**Comment by ctubbsii** [comment]
I have the start of a fix, but fixing this gets down into the weeds of how we manage thread pools.

**Comment by dlmarion** [comment]
@Manno15  @milleruntime  - I stood up an Accumulo instance from the current HEAD of `main`. I created a table, added some data, and compacted it and saw no error.

```
insert a b c d
flush
2021-11-30T18:38:10,220 [shell.Shell] INFO : Flush of table compaction_test  initiated...
insert a b d e
flush
2021-11-30T18:38:14,561 [shell.Shell] INFO : Flush of table compaction_test  initiated...
insert a d b c
flush
2021-11-30T18:38:22,226 [shell.Shell] INFO : Flush of table compaction_test  initiated...
insert a g e f 
flush
2021-11-30T18:38:31,571 [shell.Shell] INFO : Flush of table compaction_test  initiated...
compact -w 
2021-11-30T18:38:33,563 [shell.Shell] INFO : Compacting table ...
2021-11-30T18:38:34,375 [shell.Shell] INFO : Compaction of table compaction_test completed for given range
```

Is this still an issue?

**Comment by dlmarion** [comment]
> Create a table utilizing the new compaction server props

Maybe I didn't do this. I think it would definitely be useful to include as much specific detailed information as possible, configuration settings, logs, etc. It would be a big help in trying to replicate the issue.

**Comment by Manno15** [comment]
@dlmarion I still see this occurring on my end. I used the properties set in accumulo-testing for creating the table. Possible the issue is related to that. I will paste below what is in accumulo-testing.

```
tserver.compaction.major.service.cs1.planner=org.apache.accumulo.core.spi.compaction.DefaultCompactionPlanner \
tserver.compaction.major.service.cs1.planner.opts.executors=\
[{"name":"small","type":"internal","maxSize":"16M","numThreads":8},\
{"name":"medium","type":"internal","maxSize":"128M","numThreads":4},\
{"name":"large","type":"internal","numThreads":2}]
```

**Comment by dlmarion** [comment]
So, I set the following properties:

```
site       | tserver.compaction.major.service.cs1.planner ..................... | 
system     |    @override ..................................................... | org.apache.accumulo.core.spi.compaction.DefaultCompactionPlanner
site       | tserver.compaction.major.service.cs1.planner.opts.executors ...... | 
system     |    @override ..................................................... | [{"name":"small","type":"internal","maxSize":"16M","numThreads":8},{"name":"medium","type":"internal","maxSize":"128M","numThreads":4},{"name":"large","type":"internal","numThreads":2}]
```

and set the following properties on the table:
```
default    | table.compaction.dispatcher ............... | org.apache.accumulo.core.spi.compaction.SimpleCompactionDispatcher
table      | table.compaction.dispatcher.opts.service .. | cs1
```

and I'm not seeing the error.

**Comment by dlmarion** [comment]
I still don't know what the error is. if it's still happening, can you drop the stack trace here?

**Comment by Manno15** [comment]
The server-side logs are in the details portion of the issue body. I will see if I can unearth any other helpful logs. I will try to reproduce it manually but if that cannot be done then its possible the issue resides in how we set the properties in accumulo-testing.

**Comment by Manno15** [comment]
@dlmarion I did get it to fail when setting the properties manually but it took a lot longer. It didn't seem to fail until after I lowered the split threshold to cause some splits.

**Comment by ctubbsii** [comment]
@dlmarion I'm working on this. Holiday interrupted it, but the basic problem is that we are making certain assumptions about the Runnables that we put in the executor blocking queues. There are two kinds of problems: 1) we conditionally remove them from the queue, possibly based on equality, and 2) we cast these Runnables to a specific subtype to perform actions on them (sorting, removal, etc.). Both these kinds of operations are broken when we wrap the Runnables with the tracing context lambda Runnable. In some cases, we throw an IllegalArgumentException when the `instanceof` type doesn't match what we expect, and in other cases the operation is just going to fail silently because the condition doesn't match (either the objects don't equal or the `instanceof` type doesn't match).

**Comment by dlmarion** [comment]
@ctubbsii - ok. I started working on a solution, but I'll move on to something else.

**Comment by dlmarion** [comment]
[wip](https://github.com/dlmarion/accumulo/commit/c2073c5046ddc5b9bd5061f9f9265aefac4d98aa) - in case you were wondering where I was headed

**Comment by ctubbsii** [comment]
> [wip](https://github.com/dlmarion/accumulo/commit/c2073c5046ddc5b9bd5061f9f9265aefac4d98aa) - in case you were wondering where I was headed

That's the basic direction I was headed also. I called mine `TraceWrapped{Callable,Runnable}`, but had the same `unwrap` method. The main thing I did differently was I saved a reference to the original context, to call make it current in the `run` method, rather than use the `Context.wrap` method at all. That way, it's only wrapped once, instead of twice. I was going through the rest of the use cases to see if there's any further cleanup I can do to make sure there aren't similar issues. Another issue that is possible to encounter is trying to remove a wrapped runnable from the blocking queue, and not actually removing it because the wrapper doesn't equal the wrapped Runnable or isn't the right type.

**Comment by milleruntime** [comment]
Any progress on this?

**Comment by ctubbsii** [comment]
> Any progress on this?

Still working on it. Sorry.

**Comment by dlmarion** [comment]
@ctubbsii  - any update on this? Do you need any help?

**Comment by ctubbsii** [comment]
> any update on this? Do you need any help?

No update. I'm still working on it. No, I don't need help. I just need to cut out distractions to wrap it up.

**Comment by dlmarion** [comment]
Just a note to say that we have two cases where using external compactions when testing 2.1.0-SNAPSHOT does not exhibit this error. It seems only internal compactions are affected.

**Comment by keith-turner** [comment]
I was running a test with a slightly older version of 2.1.0-SNAPSHOT w/o #2442 changes.  A tablet ran into the following error when trying to split and it left the tablet in a really bad state.  I think this is a different manifestation of this bug and I am hoping that pulling in #2442 will fix my problem.  @ctubbsii curious if you think this is the same bug and your changes in #2442 will avoid the problem?  If your not sure, don't worry about it I will be spinning up a new test w/ latest code.

```
2022-02-09T18:40:20,070 [tserver.TabletServer] ERROR: Unknown error on split:
java.lang.IllegalArgumentException: Unknown runnable type io.opentelemetry.context.Context$$Lambda$376/0x0000000840441040
        at org.apache.accumulo.tserver.compactions.InternalCompactionExecutor$InternalJob.lambda$cancel$0(InternalCompactionExecutor.java:134) ~[accumulo-tserver-2.1.0-SNAPSHOT.jar:2.1.0-SNAPSHOT]
        at java.util.concurrent.PriorityBlockingQueue.bulkRemove(PriorityBlockingQueue.java:1064) ~[?:?]
        at java.util.concurrent.PriorityBlockingQueue.removeIf(PriorityBlockingQueue.java:1024) ~[?:?]
        at org.apache.accumulo.tserver.compactions.InternalCompactionExecutor$InternalJob.cancel(InternalCompactionExecutor.java:128) ~[accumulo-tserver-2.1.0-SNAPSHOT.jar:2.1.0-SNAPSHOT]
        at org.apache.accumulo.tserver.compactions.InternalCompactionExecutor.lambda$compactableClosed$4(InternalCompactionExecutor.java:221) ~[accumulo-tserver-2.1.0-SNAPSHOT.jar:2.1.0-SNAPSHOT]
        at java.util.ArrayList.forEach(ArrayList.java:1541) ~[?:?]
        at org.apache.accumulo.tserver.compactions.InternalCompactionExecutor.compactableClosed(InternalCompactionExecutor.java:221) ~[accumulo-tserver-2.1.0-SNAPSHOT.jar:2.1.0-SNAPSHOT]
        at org.apache.accumulo.tserver.compactions.CompactionService.lambda$compactableClosed$16(CompactionService.java:426) ~[accumulo-tserver-2.1.0-SNAPSHOT.jar:2.1.0-SNAPSHOT]
        at java.lang.Iterable.forEach(Iterable.java:75) ~[?:?]
        at org.apache.accumulo.tserver.compactions.CompactionService.compactableClosed(CompactionService.java:426) ~[accumulo-tserver-2.1.0-SNAPSHOT.jar:2.1.0-SNAPSHOT]
        at org.apache.accumulo.tserver.compactions.CompactionManager.lambda$compactableClosed$13(CompactionManager.java:413) ~[accumulo-tserver-2.1.0-SNAPSHOT.jar:2.1.0-SNAPSHOT]
        at java.util.stream.ForEachOps$ForEachOp$OfRef.accept(ForEachOps.java:183) ~[?:?]
        at java.util.stream.ReferencePipeline$2$1.accept(ReferencePipeline.java:177) ~[?:?]
        at java.util.stream.ReferencePipeline$3$1.accept(ReferencePipeline.java:195) ~[?:?]
        at java.util.concurrent.ConcurrentSkipListMap$KeySpliterator.forEachRemaining(ConcurrentSkipListMap.java:3156) ~[?:?]
        at java.util.stream.AbstractPipeline.copyInto(AbstractPipeline.java:484) ~[?:?]
        at java.util.stream.AbstractPipeline.wrapAndCopyInto(AbstractPipeline.java:474) ~[?:?]
        at java.util.stream.ForEachOps$ForEachOp.evaluateSequential(ForEachOps.java:150) ~[?:?]
        at java.util.stream.ForEachOps$ForEachOp$OfRef.evaluateSequential(ForEachOps.java:173) ~[?:?]
        at java.util.stream.AbstractPipeline.evaluate(AbstractPipeline.java:234) ~[?:?]
        at java.util.stream.ReferencePipeline.forEach(ReferencePipeline.java:497) ~[?:?]
        at org.apache.accumulo.tserver.compactions.CompactionManager.compactableClosed(CompactionManager.java:413) ~[accumulo-tserver-2.1.0-SNAPSHOT.jar:2.1.0-SNAPSHOT]
        at org.apache.accumulo.tserver.tablet.CompactableImpl.close(CompactableImpl.java:1517) ~[accumulo-tserver-2.1.0-SNAPSHOT.jar:2.1.0-SNAPSHOT]
        at org.apache.accumulo.tserver.tablet.Tablet.initiateClose(Tablet.java:1244) ~[accumulo-tserver-2.1.0-SNAPSHOT.jar:2.1.0-SNAPSHOT]
        at org.apache.accumulo.tserver.tablet.Tablet.split(Tablet.java:1702) ~[accumulo-tserver-2.1.0-SNAPSHOT.jar:2.1.0-SNAPSHOT]
        at org.apache.accumulo.tserver.TabletServer.splitTablet(TabletServer.java:483) ~[accumulo-tserver-2.1.0-SNAPSHOT.jar:2.1.0-SNAPSHOT]
        at org.apache.accumulo.tserver.TabletServer.splitTablet(TabletServer.java:470) ~[accumulo-tserver-2.1.0-SNAPSHOT.jar:2.1.0-SNAPSHOT]
        at org.apache.accumulo.tserver.TabletServer$SplitRunner.run(TabletServer.java:411) ~[accumulo-tserver-2.1.0-SNAPSHOT.jar:2.1.0-SNAPSHOT]
        at io.opentelemetry.context.Context.lambda$wrap$1(Context.java:207) ~[opentelemetry-context-1.7.1.jar:1.7.1]
        at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1128) ~[?:?]
        at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:628) ~[?:?]
        at io.opentelemetry.context.Context.lambda$wrap$1(Context.java:207) ~[opentelemetry-context-1.7.1.jar:1.7.1]
        at java.lang.Thread.run(Thread.java:829) [?:?]
```

**Comment by ctubbsii** [comment]
> I was running a test with a slightly older version of 2.1.0-SNAPSHOT w/o #2442 changes. A tablet ran into the following error when trying to split and it left the tablet in a really bad state. I think this is a different manifestation of this bug and I am hoping that pulling in #2442 will fix my problem. @ctubbsii curious if you think this is the same bug and your changes in #2442 will avoid the problem? If your not sure, don't worry about it I will be spinning up a new test w/ latest code.
> 
> ```
> 2022-02-09T18:40:20,070 [tserver.TabletServer] ERROR: Unknown error on split:
> java.lang.IllegalArgumentException: Unknown runnable type io.opentelemetry.context.Context$$Lambda$376/0x0000000840441040
>         at org.apache.accumulo.tserver.compactions.InternalCompactionExecutor$InternalJob.lambda$cancel$0(InternalCompactionExecutor.java:134) ~[accumulo-tserver-2.1.0-SNAPSHOT.jar:2.1.0-SNAPSHOT]
>         at java.util.concurrent.PriorityBlockingQueue.bulkRemove(PriorityBlockingQueue.java:1064) ~[?:?]
>         at java.util.concurrent.PriorityBlockingQueue.removeIf(PriorityBlockingQueue.java:1024) ~[?:?]
> ```

I'm not 100% certain, but I do believe it is another manifestation that was fixed in #2442 . Specifically, note the changes around the `removeIf`. I remember that the problem wasn't just with the comparators unwrapping things, but also any time we tried to directly manipulate the blocking queues and we expected a type any more specific than "Runnable".
